components/server/src/workspace/workspace-service.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { inject, injectable } from "inversify";
+import { WorkspaceDB } from "@gitpod/gitpod-db/lib";
+import { Project, User, Workspace, WorkspaceContext, WorkspaceSoftDeletion } from "@gitpod/gitpod-protocol";
+import { ErrorCodes, ApplicationError } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { Authorizer } from "../authorization/authorizer";
+import { TraceContext } from "@gitpod/gitpod-protocol/lib/util/tracing";
+import { WorkspaceFactory } from "./workspace-factory";
+import { StopWorkspacePolicy } from "@gitpod/ws-manager/lib";
+import { WorkspaceStarter } from "./workspace-starter";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+
+@injectable()
+export class WorkspaceService {
+    constructor(
+        @inject(WorkspaceFactory) private readonly factory: WorkspaceFactory,
+        @inject(WorkspaceStarter) private readonly workspaceStarter: WorkspaceStarter,
+        @inject(WorkspaceDB) private readonly db: WorkspaceDB,
+        @inject(Authorizer) private readonly auth: Authorizer,
+    ) {}
+
+    async createWorkspace(
+        ctx: TraceContext,
+        user: User,
+        organizationId: string,
+        project: Project | undefined,
+        context: WorkspaceContext,
+        normalizedContextURL: string,
+    ): Promise<Workspace> {
+        await this.auth.checkPermissionOnOrganization(user.id, "create_workspace", organizationId);
+
+        // We don't want to be doing this in a transaction, because it calls out to external systems.
+        // TODO(gpl) Would be great to sepearate workspace creation from external calls
+        const workspace = await this.factory.createForContext(
+            ctx,
+            user,
+            organizationId,
+            project,
+            context,
+            normalizedContextURL,
+        );
+
+        // Instead, we fall back to removing access in case something goes wrong.
+        try {
+            await this.auth.createWorkspaceInOrg(organizationId, user.id, workspace.id);
+        } catch (err) {
+            await this.hardDeleteWorkspace(user.id, workspace.id).catch((err) =>
+                log.error("failed to hard-delete workspace", err),
+            );
+            throw err;
+        }
+
+        return workspace;
+    }
+
+    async getWorkspace(userId: string, workspaceId: string): Promise<Workspace> {
+        await this.auth.checkPermissionOnWorkspace(userId, "access", workspaceId);
+
+        const workspace = await this.db.findById(workspaceId);
+        if (!workspace || !!workspace.softDeleted || workspace.deleted) {
```

Comment:
I would have been more concerned about the contract to throw NOT_FOUND instead of `undefined` but since you only replaced usages of `GitpodServerImpl#internalGetWorkspace` which has the same contract I believe this is good. 
That said, we will want to call this for `adminGetWorkspace(s)` as well, which should return deleted workspaces.

Reply:
Yes, I thought I can postpone that problem, but that might not mamke sense. Let me check if we can remove softDeleted here.

Reply:
@svenefftinge Could you shoot me a link to an example? I can't find one on main atm, and wonder if you meant setting all userId types to `string | undefined`? :thinking:

Reply:
I checked again, and before, only startWorkspace checked for softDeleted. To keep the change in this PR as low as possible, I removed that check here.
I would leave it and monitor if I was there on Monday, but as I'm not, I really don't want to add extra noise to the point of these changes (FGA). :+1: